### ISSUE TITLE
Tracking Issue: Fn::ToJsonString RFC

### ISSUE THREAD

**Issue by mluk-aws** [aws-cloudformation/cfn-language-discussion#14]
## Description

Provide an intrinsic function that converts an input template block into an escaped JSON string.

Original Issue: aws-cloudformation/cfn-language-discussion#53

## Workflow

- [x] Tracking issue created (label: `status/proposed`)
- [x] RFC pull request submitted
- [x] Reviewers assigned (label: `status/review`)
- [x] Reviewers signed-off (label `approved` applied to pull request)
- [x] Final comments period (label: `status/final-comments-period`)
- [x] Approved and merged (label: `status/approved`)
- [x] Implementation started (label: `status/implementing`)
- [x] Implementation complete (label: `status/done`)
---

> Author is responsible to progress the RFC according to this checklist, and
apply the relevant labels to this issue so that the RFC table in README gets
updated.

**Comment by jiemying** [comment]
Hi have cases say it is not working at the moment

      EventPattern:
        Fn::ToJsonString:
          - test1: 'test2'
            test3: 'test4'
            
Stack will fail at the Transform stage with error "Transform AWS::LanguageExtensions failed with: Fn::ToJsonString resolve value type mismatch. Rollback requested by user."

**Comment by MalikAtalla-AWS** [comment]
Hey @jiemying,
thanks for bringing this up. 

To unblock you: [EventPattern](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html#cfn-events-rule-eventpattern) supports Json objects natively. There is no need to use Fn::ToJsonString for the EventPattern attribute.

**Comment by jiemying** [comment]
> Hey @jiemying, thanks for bringing this up.
> 
> To unblock you: [EventPattern](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html#cfn-events-rule-eventpattern) supports Json objects natively. There is no need to use Fn::ToJsonString for the EventPattern attribute.

Hi, thx for coming back to me so promptly!

Yep I am aware of that, that was for reference. 

The error happens at the Transform stage of the CloudFormation stack operation, no matter where I put the above test Fn::ToJsonString in the template's resources properties, the stack transform stage failed with the same error.

I suspect the object "type" passed to this intrinsic function in the back-end, does not match wanted object type of the code, hence the exception error throws. Look at the exception part of the code may be helpful.

**Comment by MalikAtalla-AWS** [comment]
Hey @jiemying,
if this happens even when you put Fn::ToJsonString somewhere else (where a String is expected), I would appreciate if you could provide an example template. If you still have it handy, any chance you could open a new issue in this same repo?

**Comment by lejiati** [comment]
Close this issue as it's [available](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ToJsonString.html) via `AWS::LanguageExtensions`
